File: backend/models/collab.py
"""
backend/models/collab.py
Collaboration draft models: minimal, idempotent design.
"""

from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime
from enum import Enum
from typing import Optional, List, Dict, Any


class DraftStatus(str, Enum):
    """Draft lifecycle: created -> adding_segments -> ready_to_post -> posted"""

    ACTIVE = "active"
    LOCKED = "locked"  # Ring holder passed, awaiting next action
    COMPLETED = "completed"


class DraftSegment(BaseModel):
    """Individual segment contributed to draft"""

    model_config = ConfigDict(frozen=True)

    segment_id: str = Field(description="UUID")
    draft_id: str = Field(description="Parent draft UUID")
    user_id: str = Field(description="Segment author user ID")
    content: str = Field(min_length=1, max_length=500)
    created_at: datetime
    segment_order: int = Field(description="Order in thread (0-indexed)")
    idempotency_key: Optional[str] = None
    # Phase 3.3a: Attribution fields
    author_user_id: Optional[str] = Field(default=None, description="Author user ID (same as user_id for now)")
    author_display: Optional[str] = Field(default=None, description="Deterministic display name")
    ring_holder_user_id_at_write: Optional[str] = Field(default=None, description="Ring holder when segment was written")
    ring_holder_display_at_write: Optional[str] = Field(default=None, description="Ring holder display name at write")


class RingState(BaseModel):
    """Ring holder state within draft"""

    model_config = ConfigDict(frozen=True)

    draft_id: str
    current_holder_id: str = Field(description="User ID who can append")
    holders_history: List[str] = Field(default_factory=list, description="All past holders")
    passed_at: datetime
    last_passed_at: Optional[datetime] = Field(default=None, description="ISO timestamp of last ring pass (Phase 3.3a)")
    idempotency_key: Optional[str] = None


class CollabDraft(BaseModel):
    """Collaboration draft: multiple contributors, pass-the-ring ownership"""

    model_config = ConfigDict(frozen=True)

    draft_id: str = Field(description="UUID")
    creator_id: str
    title: str = Field(max_length=200)
    platform: str = Field(
        description="Target platform: 'x', 'instagram', 'tiktok', 'youtube'"
    )
    status: DraftStatus = Field(default=DraftStatus.ACTIVE)
    segments: List[DraftSegment] = Field(default_factory=list)
    ring_state: RingState
    collaborators: List[str] = Field(
        default_factory=list, description="User IDs of accepted collaborators (excludes creator)"
    )
    pending_invites: List[str] = Field(
        default_factory=list, description="Invite IDs pending acceptance"
    )
    created_at: datetime
    updated_at: datetime
    target_publish_at: Optional[datetime] = None
    # Phase 3.3a: Ring velocity metrics (computed on read)
    metrics: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Ring velocity metrics: contributorsCount, ringPassesLast24h, avgMinutesBetweenPasses, lastActivityAt"
    )


class CollabDraftRequest(BaseModel):
    """Request to create new draft"""

    title: str = Field(min_length=1, max_length=200)
    platform: str = Field(
        description="'x' | 'instagram' | 'tiktok' | 'youtube'"
    )
    initial_segment: Optional[str] = Field(
        default=None, max_length=500, description="Optional first segment"
    )


class SegmentAppendRequest(BaseModel):
    """Request to append segment (idempotent)"""

    content: str = Field(min_length=1, max_length=500)
    idempotency_key: str = Field(
        description="UUID, prevents duplicate appends"
    )


class RingPassRequest(BaseModel):
    """Request to pass ring (idempotent)"""

    to_user_id: str
    idempotency_key: str = Field(
        description="UUID, prevents duplicate passes"
    )

File: src/types/collab.ts
/**
 * Collaboration types for frontend.
 * Mirrors backend models in backend/models/collab.py
 */

export type DraftStatus = "active" | "locked" | "completed";

export interface DraftSegment {
  segment_id: string;
  draft_id: string;
  user_id: string;
  content: string;
  created_at: string;
  segment_order: number;
  idempotency_key?: string;
  author_user_id?: string;
  author_display?: string;
  ring_holder_user_id_at_write?: string;
  ring_holder_display_at_write?: string;
}

export interface RingState {
  draft_id: string;
  current_holder_id: string;
  holders_history: string[];
  passed_at: string;
  last_passed_at?: string;
  idempotency_key?: string;
}

export interface CollabDraft {
  draft_id: string;
  creator_id: string;
  title: string;
  platform: string;
  status: DraftStatus;
  segments: DraftSegment[];
  ring_state: RingState;
  collaborators: string[];
  pending_invites: string[];
  created_at: string;
  updated_at: string;
  target_publish_at?: string;
  metrics?: {
    contributorsCount?: number;
    ringPassesLast24h?: number;
    avgMinutesBetweenPasses?: number;
    lastActivityAt?: string;
  };
}

export interface CollabDraftRequest {
  title: string;
  platform: string;
  initial_segment?: string;
}

export interface SegmentAppendRequest {
  content: string;
  idempotency_key: string;
}

export interface RingPassRequest {
  to_user_id: string;
  idempotency_key: string;
}

export interface AISuggestion {
  mode: "next" | "rewrite" | "summary" | "commentary";
  content: string;
  ring_holder: boolean;
  platform?: string;
  generated_at?: string;
}

export type FormatPlatform = "x" | "youtube" | "instagram" | "blog";

export interface FormatBlock {
  type: "heading" | "text" | "hashtag" | "cta" | "media_note";
  text: string;
  heading?: string;
}

export interface PlatformOutput {
  platform: FormatPlatform;
  blocks: FormatBlock[];
  plain_text: string;
  character_count: number;
  block_count: number;
  warnings: string[];
}

export interface FormatOptions {
  tone?: "professional" | "casual" | "witty" | "motivational" | "technical";
  include_hashtags?: boolean;
  include_cta?: boolean;
  hashtag_count?: number;
  hashtag_suggestions?: string[];
  cta_text?: string;
  cta_suggestions?: string[];
}

export interface FormatGenerateRequest {
  draft_id: string;
  platforms?: FormatPlatform[];
  options?: FormatOptions;
}

export interface FormatGenerateResponse {
  draft_id: string;
  outputs: Record<FormatPlatform, PlatformOutput>;
}

// Timeline types (Phase 8.3)
export type TimelineEventType = 
  | "draft_created"
  | "segment_added"
  | "ring_passed"
  | "collaborator_added"
  | "ai_suggested"
  | "format_generated"
  | "other";

export interface TimelineEvent {
  event_id: string;
  ts: string;  // ISO datetime
  type: TimelineEventType;
  actor_user_id: string | null;
  draft_id: string;
  summary: string;
  meta: Record<string, any>;
}

export interface TimelineResponse {
  draft_id: string;
  events: TimelineEvent[];
  next_cursor?: string;
}

export interface ContributorStats {
  user_id: string;
  segment_count: number;
  segment_ids: string[];
  first_ts: string;
  last_ts: string;
}

export interface AttributionResponse {
  draft_id: string;
  contributors: ContributorStats[];
}

export interface ExportRequest {
  format: "markdown" | "json";
  include_credits?: boolean;
}

export interface ExportResponse {
  draft_id: string;
  format: string;
  filename: string;
  content_type: string;
  content: string;
}

export type ErrorCode = 
  | "ring_required"
  | "rate_limited"
  | "permission_denied"
  | "not_found"
  | "validation_error"
  | "unknown_error";

export interface APIError {
  code: ErrorCode;
  message: string;
  status: number;
}
